common/src/main/java/com/lightstep/tracer/shared/AbstractTracer.java

```diff
@@ -346,19 +346,30 @@ public Boolean flush(long timeoutMillis) {
 
     protected abstract SimpleFuture<Boolean> flushInternal(boolean explicitRequest);
 
+    /**
+     * Do not call this function while holding a lock.
+     *
+     */
     private boolean initializeCollectorClient() {
+        CollectorClient newClient;
         try {
             ManagedChannelBuilder builder =
                 ManagedChannelBuilder.forAddress(collectorURL.getHost(), collectorURL.getPort());
             if (collectorURL.getProtocol() == "http") {
                 builder.usePlaintext(true);
             }
-            client = new CollectorClient(this, builder);
```

Comment:
Where do we now call initializeCollectorClient?

I think since we call initiazlieCollectorClient lazily, we actually *do* need synchronization around calling it, and possibly around accessing the client variable. Should we instead initialize it at setup so we don't have to coordinate these things?